### PR TITLE
Detect when the wrong type of cpp toolchain is loaded.

### PR DESCRIPTION
**Pull request by katre**
Change-Id: I8aa268af62f187268b3da3740ba1ada57c69f34c

### CODE REVIEW

File: src/main/java/com/google/devtools/build/lib/rules/cpp/CppHelper.java [hunk 1]
```diff
@@ -518,7 +518,13 @@ private static CcToolchainProvider getToolchainFromPlatformConstraints(
...
         (ResolvedToolchainProviders)
             ruleContext.getToolchainContext().getResolvedToolchainProviders();
-    return (CcToolchainProvider) providers.getForToolchainType(toolchainType);
+    ToolchainInfo toolchainInfo = providers.getForToolchainType(toolchainType);
+    if (toolchainInfo instanceof CcToolchainProvider) {
+      return (CcToolchainProvider) toolchainInfo;
+    }
+
+    ruleContext.ruleError("The selected C++ toolchain is not a cc_toolchain rule");
+    return CcToolchainProvider.EMPTY_TOOLCHAIN_IS_ERROR;
```
Comment (lberki) [inline src/main/java/com/google/devtools/build/lib/rules/cpp/CppHelper.java]: This looks good per se, but wouldn't this be a problem with other toolchain types? Are all users of toolchains expected to check that they got the kind of toolchain that they expect?

File: src/main/java/com/google/devtools/build/lib/rules/cpp/CppHelper.java [hunk 2]
```diff
@@ -518,7 +518,13 @@ private static CcToolchainProvider getToolchainFromPlatformConstraints(
     ResolvedToolchainProviders providers =
         (ResolvedToolchainProviders)
             ruleContext.getToolchainContext().getResolvedToolchainProviders();
-    return (CcToolchainProvider) providers.getForToolchainType(toolchainType);
+    ToolchainInfo toolchainInfo = providers.getForToolchainType(toolchainType);
+    if (toolchainInfo instanceof CcToolchainProvider) {
+      return (CcToolchainProvider) toolchainInfo;
+    }
+
+    ruleContext.ruleError("The selected C++ toolchain is not a cc_toolchain rule");
```
Comment (lberki) [inline src/main/java/com/google/devtools/build/lib/rules/cpp/CppHelper.java]: Can I haz test case?

File: src/main/java/com/google/devtools/build/lib/rules/cpp/CppHelper.java (hunk 1 above)
Comment (katre) [reply src/main/java/com/google/devtools/build/lib/rules/cpp/CppHelper.java]: In general, users of toolchains should make this check. This is a special case, in that a dummy toolchain was added to make toolchain resolution happy, before the cc_toolchain was fixed.

In general, Skylark rules will all be using ToolchainInfo, which removes the issue with data casts, but raises the issue that ToolchainInfo is basically just a bag of data, and if toolchains are configured improperly, the wrong bag will arrive. I'd like to come up with a general solution for this, but I'm not sure what that would look like.

**Comment by katre** [comment]
I added a commit changing all callers of getToolchain to handle the RuleErrorException, let me know what you think.

**Comment by lberki** [comment]
Yuck. Better than crashing, I guess, but those "throws RuleErrorException" clauses are painful. My brain is already fried for the day and GitHub is not helping, but wouldn't it be easier to call `getToolchainFromPlatformConstraints` earlier and thus have to thread this exception through less places? But then again, then you have to thread `CcToolchain`, which isn't a lot of improvement.

Any alternatives? I can't think of anything due to the aforementioned fried brain, so if you can't come up with anything until tomorrow, I guess this'll have to do for lack of a better idea. At least this just makes the code ugly and won't require a migration later...

**Comment by katre** [comment]
I can't come up with a better solution. If we accept that calling getToolchain can sometimes fail (which it could before even in the non-platforms case, but it led to a NPE later on), we need the RuleErrorExceptions.

**Comment by lberki** [comment]
Let's break this out into an e-mail thread. This is a lot of collateral damage and I don't have any great ideas, either (other than resorting to unchecked exceptions, which can hardly be called "great"...)

**Pull request closed without merging**